sdks/python/apache_beam/runners/portability/portable_metrics.py

```diff
@@ -22,53 +22,44 @@
-  @staticmethod
-  def from_monitoring_infos(monitoring_info_list, user_metrics_only=False):
-    """Groups MonitoringInfo objects into counters, distributions and gauges.
+def from_monitoring_infos(monitoring_info_list, user_metrics_only=False):
+  """Groups MonitoringInfo objects into counters, distributions and gauges.
 
-    Args:
-      monitoring_info_list: An iterable of MonitoringInfo objects.
-      user_metrics_only: If true, includes user metrics only.
-    """
-    counters = {}
-    distributions = {}
-    gauges = {}
+  Args:
+    monitoring_info_list: An iterable of MonitoringInfo objects.
+    user_metrics_only: If true, includes user metrics only.
+  Returns:
+    A tuple containing three dictionaries: counters, distributions and gauges,
+    respectively. Each dictionary contains (MetricKey, metric result) pairs.
+  """
+  counters = {}
+  distributions = {}
+  gauges = {}
 
-    for mi in monitoring_info_list:
-      if (user_metrics_only and
-          not monitoring_infos.is_user_monitoring_info(mi)):
-        continue
+  for mi in monitoring_info_list:
+    if (user_metrics_only and
+        not monitoring_infos.is_user_monitoring_info(mi)):
+      continue
 
-      key = ParseMonitoringInfoMixin._create_metric_key(mi)
-      metric_result = (monitoring_infos.extract_metric_result_map_value(mi))
+    try:
+      key = _create_metric_key(mi)
+    except ValueError:
+      continue
+    metric_result = (monitoring_infos.extract_metric_result_map_value(mi))
 
-      if monitoring_infos.is_counter(mi):
-        counters[key] = metric_result
-      elif monitoring_infos.is_distribution(mi):
-        distributions[key] = metric_result
-      elif monitoring_infos.is_gauge(mi):
-        gauges[key] = metric_result
+    if monitoring_infos.is_counter(mi):
+      counters[key] = metric_result
+    elif monitoring_infos.is_distribution(mi):
+      distributions[key] = metric_result
+    elif monitoring_infos.is_gauge(mi):
+      gauges[key] = metric_result
 
-    return counters, distributions, gauges
+  return counters, distributions, gauges
 
-  @staticmethod
-  def _create_metric_key(monitoring_info):
-    step_name = ParseMonitoringInfoMixin._get_step_name(monitoring_info)
-    namespace, name = monitoring_infos.parse_namespace_and_name(monitoring_info)
-    return MetricKey(step_name, MetricName(namespace, name))
 
-  @staticmethod
-  def _get_step_name(monitoring_info):
-    keys_to_check = [monitoring_infos.PTRANSFORM_LABEL,
-                     monitoring_infos.PCOLLECTION_LABEL,
-                     'WINDOWING_STRATEGY',
-                     'CODER',
-                     'ENVIRONMENT',
-                    ]
-
-    for key in keys_to_check:
-      value = monitoring_info.labels.get(key)
-      if value:
-        return value
-    return ''
+def _create_metric_key(monitoring_info):
+  step_name = monitoring_infos.get_step_name(monitoring_info)
+  if not step_name:
+    raise ValueError("Step name is empty")
```

Comment:
Dump monitoring info to error message. Otherwise this error is indescriptive. Ideally change text to what you expect here, for example: "Monitoring info should contain XXX field. Monitoring info: {}" or in this case it can be: "Failed to deduce step_name from MonitoringInfo: {}"

Reply:
"Failed to deduce step_name..." looks good.